### PR TITLE
Add `fromJson` and `mergeFromJson` and their string variants

### PR DESCRIPTION
**Pull request by srikrsna-buf**
Add `fromJson` and `mergeFromJson` and their string variants. Missing extension support, will a be added in a follow up.

### CODE REVIEW

File: packages/protobuf/src/next/from-json.ts [hunk 1]
```diff
@@ -0,0 +1,887 @@
...
+      if (typeof json !== "string") break;
+      return base64Decode(json);
+  }
+  throw new Error();
+}
+
+function findJsonName(msg: ReflectMessage, name: string) {
+  for (const f of msg.fields) {
+    // eslint-disable-next-line @typescript-eslint/prefer-nullish-coalescing
+    const jsonName = f.jsonName! || protoCamelCase(f.name);
```
[the remark this answers is not in the document]
Comment (srikrsna-buf) [reply packages/protobuf/src/next/from-json.ts:627]: This is to handle the buf where it is not undefined and currently return an empty string.

[the remark this answers is not in the document]
Comment (timostamm) [reply packages/protobuf/src/next/from-json.ts:627]: The bug was fixed in https://github.com/bufbuild/protobuf-es/commit/3fddb408a00b8df8eb624b299b41c2b3a92397ec.

File: packages/protobuf/src/next/from-json.ts [hunk 2]
```diff
@@ -0,0 +1,887 @@
...
+
+export function fromJson<Desc extends DescMessage>(
+  desc: Desc,
+  targetOrJson: MessageShape<Desc> | JsonValue,
+  jsonOrOptions?: JsonValue | Partial<JsonReadOptions>,
+  options?: Partial<JsonReadOptions>,
+): MessageShape<Desc> {
+  let msg: ReflectMessage;
+  let json: JsonValue;
+  if (isMessage(targetOrJson)) {
```
[the remark this answers is not in the document]
Comment (srikrsna-buf) [reply packages/protobuf/src/next/from-json.ts:177]: Added a comment for now, I'll take this up immediately after this as it touches the signature of everything.

[the remark this answers is not in the document]
Comment (srikrsna-buf) [reply packages/protobuf/src/next/from-json.ts:177]: Scratch that updated to use merge functions, opened a new PR for binary in #772

**Review by timostamm — APPROVED** [review]
Nice, LGTM!

After we have good tests coverage, we need to look into:
- de-duplicate checks in from-json.ts and reflect.ts
- provide better error messages in `fromJson()`


For context regarding error messages: The v1 runtime's error messages are not very helpful. `FieldInfo` does not have access to parents, and cannot provide the full path to an invalid field. This has changes with the v2 runtime, reflect.ts is making use of it, and our parsers should as well.

**Pull request merged**